File: src/apps/daemon/internal/platform/security/wireguard_ipam_manager.go
package security

import (
	"errors"
	"fmt"
	"strings"
)

type WireguardPeer struct {
	PublicKey  string   `json:"public_key"`
	AssignedIP string   `json:"assigned_ip"`
	AllowedIPs []string `json:"allowed_ips"`
	Endpoint   string   `json:"endpoint"`
	Enabled    bool     `json:"enabled"`
}

type WireguardIpamManager struct {
	subnetPrefix string
	nextSuffix   uint8
	serverIP     string
	listenPort   uint16
	peers        map[string]*WireguardPeer
	ipToKey      map[string]string
}

func NewWireguardIpamManager(subnetPrefix string, listenPort uint16) (*WireguardIpamManager, error) {
	parts := strings.Split(subnetPrefix, ".")
	if len(parts) != 3 {
		return nil, errors.New("subnet prefix must have 3 octets, e.g. 10.14.0")
	}

	return &WireguardIpamManager{
		subnetPrefix: subnetPrefix,
		nextSuffix:   2,
		serverIP:     fmt.Sprintf("%s.1", subnetPrefix),
		listenPort:   listenPort,
		peers:        make(map[string]*WireguardPeer),
		ipToKey:      make(map[string]string),
	}, nil
}

func (w *WireguardIpamManager) AllocatePeer(publicKey string) (*WireguardPeer, error) {
	if peer, ok := w.peers[publicKey]; ok {
		return peer, nil
	}

	if w.nextSuffix >= 254 {
		return nil, errors.New("subnet exhausted")
	}

	assignedIP := fmt.Sprintf("%s.%d", w.subnetPrefix, w.nextSuffix)
	w.nextSuffix++

	peer := &WireguardPeer{
		PublicKey:  publicKey,
		AssignedIP: assignedIP,
		AllowedIPs: []string{fmt.Sprintf("%s/32", assignedIP)},
		Enabled:    true,
	}

	w.peers[publicKey] = peer
	w.ipToKey[assignedIP] = publicKey
	return peer, nil
}

func (w *WireguardIpamManager) ReleasePeer(publicKey string) error {
	peer, ok := w.peers[publicKey]
	if !ok {
		return errors.New("peer not found")
	}

	delete(w.ipToKey, peer.AssignedIP)
	delete(w.peers, publicKey)
	return nil
}

func (w *WireguardIpamManager) GenerateServerConfig(serverPrivateKey string) string {
	var lines []string
	lines = append(lines, "[Interface]")
	lines = append(lines, fmt.Sprintf("Address = %s/24", w.serverIP))
	lines = append(lines, fmt.Sprintf("ListenPort = %d", w.listenPort))
	lines = append(lines, fmt.Sprintf("PrivateKey = %s", serverPrivateKey))
	lines = append(lines, "")

	for _, peer := range w.peers {
		if peer.Enabled {
			lines = append(lines, "[Peer]")
			lines = append(lines, fmt.Sprintf("PublicKey = %s", peer.PublicKey))
			lines = append(lines, fmt.Sprintf("AllowedIPs = %s", strings.Join(peer.AllowedIPs, ", ")))
			if peer.Endpoint != "" {
				lines = append(lines, fmt.Sprintf("Endpoint = %s", peer.Endpoint))
			}
			lines = append(lines, "")
		}
	}

	return strings.Join(lines, "\n")
}

func (w *WireguardIpamManager) GenerateClientConfig(peerPublicKey, peerPrivateKey, serverPublicKey, serverEndpoint string) (string, error) {
	peer, ok := w.peers[peerPublicKey]
	if !ok {
		return "", errors.New("peer not registered")
	}

	var lines []string
	lines = append(lines, "[Interface]")
	lines = append(lines, fmt.Sprintf("Address = %s/32", peer.AssignedIP))
	lines = append(lines, fmt.Sprintf("PrivateKey = %s", peerPrivateKey))
	lines = append(lines, "DNS = 1.1.1.1")
	lines = append(lines, "")

	lines = append(lines, "[Peer]")
	lines = append(lines, fmt.Sprintf("PublicKey = %s", serverPublicKey))
	lines = append(lines, fmt.Sprintf("Endpoint = %s", serverEndpoint))
	lines = append(lines, "AllowedIPs = 0.0.0.0/0, ::/0")
	lines = append(lines, "PersistentKeepalive = 25")

	return strings.Join(lines, "\n"), nil
}

File: src/packages/lumicore/src/security/wireguard_ipam_manager.rs
//! # WireGuard IPAM Manager
//!
//! WireGuard IP address management (IPAM), peer key registry, dynamic subnet allocator,
//! and server/client interface configuration synthesizer.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WireguardPeer {
    pub public_key: String,
    pub assigned_ip: String,
    pub allowed_ips: Vec<String>,
    pub endpoint: Option<String>,
    pub enabled: bool,
}

pub struct WireguardIpamManager {
    subnet_prefix: String,
    next_ip_suffix: u8,
    server_ip: String,
    listen_port: u16,
    peers: HashMap<String, WireguardPeer>, // public_key -> peer
    ip_to_key: HashMap<String, String>,
}

impl WireguardIpamManager {
    pub fn new(subnet_prefix: &str, listen_port: u16) -> Result<Self, String> {
        let parts: Vec<&str> = subnet_prefix.split('.').collect();
        if parts.len() != 3 {
            return Err("Subnet prefix must have 3 octets, e.g. 10.14.0".to_string());
        }

        let server_ip = format!("{}.1", subnet_prefix);
        Ok(Self {
            subnet_prefix: subnet_prefix.to_string(),
            next_ip_suffix: 2,
            server_ip,
            listen_port,
            peers: HashMap::new(),
            ip_to_key: HashMap::new(),
        })
    }

    pub fn allocate_peer(&mut self, public_key: &str) -> Result<WireguardPeer, String> {
        if let Some(existing) = self.peers.get(public_key) {
            return Ok(existing.clone());
        }

        if self.next_ip_suffix >= 254 {
            return Err("Subnet exhausted".to_string());
        }

        let assigned_ip = format!("{}.{}", self.subnet_prefix, self.next_ip_suffix);
        self.next_ip_suffix += 1;

        let peer = WireguardPeer {
            public_key: public_key.to_string(),
            assigned_ip: assigned_ip.clone(),
            allowed_ips: vec![format!("{}/32", assigned_ip)],
            endpoint: None,
            enabled: true,
        };

        self.ip_to_key.insert(assigned_ip, public_key.to_string());
        self.peers.insert(public_key.to_string(), peer.clone());
        Ok(peer)
    }

    pub fn release_peer(&mut self, public_key: &str) -> Result<(), String> {
        if let Some(peer) = self.peers.remove(public_key) {
            self.ip_to_key.remove(&peer.assigned_ip);
            Ok(())
        } else {
            Err("Peer not found".to_string())
        }
    }

    pub fn generate_server_config(&self, server_private_key: &str) -> String {
        let mut out = Vec::new();
        out.push("[Interface]".to_string());
        out.push(format!("Address = {}/24", self.server_ip));
        out.push(format!("ListenPort = {}", self.listen_port));
        out.push(format!("PrivateKey = {}", server_private_key));
        out.push("".to_string());

        for peer in self.peers.values() {
            if peer.enabled {
                out.push("[Peer]".to_string());
                out.push(format!("PublicKey = {}", peer.public_key));
                out.push(format!("AllowedIPs = {}", peer.allowed_ips.join(", ")));
                if let Some(ref ep) = peer.endpoint {
                    out.push(format!("Endpoint = {}", ep));
                }
                out.push("".to_string());
            }
        }

        out.join("\n")
    }

    pub fn generate_client_config(
        &self,
        peer_public_key: &str,
        peer_private_key: &str,
        server_public_key: &str,
        server_endpoint: &str,
    ) -> Result<String, String> {
        let peer = self.peers.get(peer_public_key).ok_or("Peer not registered")?;

        let mut out = Vec::new();
        out.push("[Interface]".to_string());
        out.push(format!("Address = {}/32", peer.assigned_ip));
        out.push(format!("PrivateKey = {}", peer_private_key));
        out.push("DNS = 1.1.1.1".to_string());
        out.push("".to_string());

        out.push("[Peer]".to_string());
        out.push(format!("PublicKey = {}", server_public_key));
        out.push(format!("Endpoint = {}", server_endpoint));
        out.push("AllowedIPs = 0.0.0.0/0, ::/0".to_string());
        out.push("PersistentKeepalive = 25".to_string());

        Ok(out.join("\n"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ipam_allocation_and_config() {
        let mut ipam = WireguardIpamManager::new("10.88.0", 51820).unwrap();
        let peer1 = ipam.allocate_peer("client_pubkey_11111111111111111111111111=").unwrap();
        assert_eq!(peer1.assigned_ip, "10.88.0.2");

        let peer2 = ipam.allocate_peer("client_pubkey_22222222222222222222222222=").unwrap();
        assert_eq!(peer2.assigned_ip, "10.88.0.3");

        let srv_cfg = ipam.generate_server_config("server_private_key_xyz=");
        assert!(srv_cfg.contains("Address = 10.88.0.1/24"));
        assert!(srv_cfg.contains("client_pubkey_11111111111111111111111111="));

        let client_cfg = ipam.generate_client_config(
            "client_pubkey_11111111111111111111111111=",
            "client_privkey_11111111111111111111111111=",
            "server_pubkey_abc=",
            "203.0.113.10:51820",
        ).unwrap();
        assert!(client_cfg.contains("Address = 10.88.0.2/32"));
        assert!(client_cfg.contains("server_pubkey_abc="));

        ipam.release_peer("client_pubkey_11111111111111111111111111=").unwrap();
        assert!(ipam.peers.get("client_pubkey_11111111111111111111111111=").is_none());
    }
}
